File: 卡码/代码随想录/1/11-图论/Js/bfs/LeetCode_0841.js
//841. 钥匙和房间
const canVisitAllRooms = function(rooms) {
  const bfs=()=>{
    const visited=new Array(rooms.length).fill(false);
    const q=new Array();
    q.push(0);
    visited[0]=true;
    while(q.length>0){
      const key=q.shift();
      const keys=rooms[key];
      for(let k of keys){
        if(!visited[k]){
          visited[k]=true;
          q.push(k);
        }
      }
    }
    return visited.every(item=>item==true);
  }
  return bfs();
};

File: 卡码/代码随想录/Java/codeThoughts/src/main/java/com/kurumi/leetCode/codeThoughts/图论/bfs/LeetCode_0841.java
package com.kurumi.leetCode.codeThoughts.图论.bfs;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 841. 钥匙和房间
 */
public class LeetCode_0841 {
    static class Solution {
        boolean bfs(List<List<Integer>> rooms) {
            boolean[] visited = new boolean[rooms.size()];
            Deque<Integer> q = new ArrayDeque<>();
            visited[0] = true;
            q.offer(0);
            while (!q.isEmpty()) {
                int key = q.poll();
                List<Integer> keys = rooms.get(key);
                for (int k : keys) {
                    if (!visited[k]) {
                        visited[k] = true;
                        q.offer(k);
                    }
                }
            }
            for (boolean visit : visited) {
                if (!visit)
                    return false;
            }
            return true;
        }

        public boolean canVisitAllRooms(List<List<Integer>> rooms) {
            return bfs(rooms);
        }
    }
}
